odoo/addons/base/models/qweb.py

```diff
@@ -397,397 +244,311 @@ def get_template(self, template, options):
+
+        code = []
+        dict_arg = []
+        for key in list(el.attrib):
+            if key.startswith('t-options-'):
+                value = el.attrib.pop(key)
+                directive += f' {key}={repr(value)}'
+                option_name = key[10:]
+                dict_arg.append(f'{repr(option_name)}:{self._compile_expr(value)}')
+
+        if t_options and dict_arg:
+            code.append(self._indent(f"{varname} = {{**{self._compile_expr(t_options)}, {', '.join(dict_arg)}}}", indent))
+        elif dict_arg:
+            code.append(self._indent(f"{varname} = {{{', '.join(dict_arg)}}}", indent))
+        elif t_options:
+            code.append(self._indent(f"{varname} = {self._compile_expr(t_options)}", indent))
+
+        return code
+
+    def _compile_format(self, expr):
+        """ Parses the provided format string and compiles it to a single
 
-        # all directives have been compiled, there should be none left
-        if any(att.startswith('t-') for att in el.attrib):
-            raise NameError("Unknown directive on %s" % etree.tostring(el, encoding='unicode'))
-        return []
+        expression python, uses string with format method.
+        Use format is faster to concat string and values.
+        """
+        text = ''
+        values = []
+        base_idx = 0
+        for m in _FORMAT_REGEX.finditer(expr):
+            literal = expr[base_idx:m.start()]
+            if literal:
+                text += literal.replace('{', '{{').replace("}", "}}")
+            text += '{}'
+            values.append(f'self._compile_to_str({self._compile_expr(m.group(1) or m.group(2))})')
+            base_idx = m.end()
+        # string past last regex match
+        literal = expr[base_idx:]
+        if literal:
+            text += literal.replace('{', '{{').replace("}", "}}")
+
+        code = repr(text)
+        if values:
+            code += f'.format({", ".join(values)})'
+        return code
+
+    def __compile_expr_tokens(self, tokens, allowed_keys, raise_on_missing=False):
+        # found allowed values
+        index = 0
+        open_bracket_index = -1
+        bracket_depth = 0
+        while index < len(tokens):
+            t = tokens[index]
+            string = t['string']
+
+            if t['type'] == token.OP:
+                if string in ['(', '[', '{']:
+                    bracket_depth += 1
+                elif string in [')', ']', '}']:
+                    bracket_depth -= 1
+            elif bracket_depth == 0 and t['type'] == token.NAME:
+                if string == 'lambda': # lambda => allowed values for the current bracket depth
+                    i = index + 1
+                    while tokens[i] and (tokens[i]['type'] == token.OP or tokens[i]['type'] == token.NAME) and tokens[i]['string'] != ':':
+                        if tokens[i]['type'] == token.NAME:
+                            allowed_keys.append(tokens[i]['string'])
+                        i += 1
```

Comment:
This (and the somewhat equivalent loop below) doesn't actually work, it lets authors smuggle globals into their code by using them as default values for arguments:
```xml
<templates>
    <t t-name="foo">
        <t t-out="lambda a=Markup: a"/>
    </t>
</templates>
```
=>
```python
def template(self, values, log):
    log["last_path_node"] = '/templates/t'
    log["last_path_node"] = '/templates/t/t'
    yield '<templates>\n    \n        '
    content = (lambda a=Markup:           a)
    force_display = None
    if content is not None and content is not False:
        yield str(escape(content))
    yield '\n    \n</templates>'
```
I'm not quite sure it gives access to important things but it seems like an oversight regardless as it makes possible to launder basically any value.

Reply:
Right, yeah, so after going through qweb-actual to make sure I wasn't looking at the base qweb's `_available_objects` (which is wildly unsafe, it should just be completely empty I think) all the builtins are accessible through this, not just the `_prepare_globals` stuff:
```xml
    <t t-set="Markup" t-value="(lambda a=Markup: a)()"/>
    <t t-set="globals" t-value="(lambda a=globals: a)()"/>
    <t t-out="Markup(str(globals()))"/>
    
    <t t-set="o" t-value="(lambda a=open: a)()"/>
    <t t-out="o('/etc/passwd').read()"/>
```
will dump the standard set of builtins (as globals) and the contents of `/etc/passwd`. The `eval` builtin is available as well. It's not clear whether it's possible to "escape" further as `_compile_expr` will prevent access to dunder attributes (e.g. `__module__`, `__self__`) but that seems academic.